lib/model/internal_link.dart

```diff
@@ -0,0 +1,285 @@
+        final suffix = element.operand.length >= 3 ? 'group' : 'pm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowDm():
+        assert(false, 'ApiNarrowDm should have been resolved');
+      case ApiNarrowMessageId():
+        fragment.write(element.operand.toString());
+    }
+  }
+
+  if (nearMessageId != null) {
+    fragment.write('/near/$nearMessageId');
+  }
+
+  return store.account.realmUrl.replace(fragment: fragment.toString());
+}
+
+/// Create a new `Uri` object in relation to a given realmUrl.
+///
+/// Returns `null` if `urlString` could not be parsed as a `Uri`.
+Uri? tryResolveToRealmUrl(String urlString, Uri realmUrl) {
+  try {
+    return realmUrl.resolve(urlString);
+  } on FormatException {
+    return null;
+  }
+}
+
+/// A [Narrow] from a given URL, on `store`'s realm.
+///
+/// Returns `null` if any of the operator/operand pairs are invalid.
+///
+/// Since narrow links can combine operators in ways our [Narrow] type can't
+/// represent, this can also return null for valid narrow links.
+///
+/// This can also return null for some valid narrow links that our Narrow
+/// type *could* accurately represent. We should try to understand these
+/// better, but some kinds will be rare, even unheard-of:
+///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
+  final url = store.account.realmUrl.resolveUri(referenceUrl);
+  if (!_isInternalLink(url, store.account.realmUrl)) return null;
+
+  final (category, segments) = _getCategoryAndSegmentsFromFragment(url.fragment);
+  switch (category) {
+    case 'narrow':
+      if (segments.isEmpty || !segments.length.isEven) return null;
+      return _interpretNarrowSegments(segments, store);
+  }
+  return null;
+}
+
+/// Check if `url` is an internal link on the given `realmUrl`.
+bool _isInternalLink(Uri url, Uri realmUrl) {
+  try {
+    if (url.origin != realmUrl.origin) return false;
+  } on StateError {
+    return false;
+  }
+  return (url.hasEmptyPath || url.path == '/')
+    && !url.hasQuery
+    && url.hasFragment;
+}
+
+/// Split `fragment` of arbitrary segments and handle trailing slashes
+(String, List<String>) _getCategoryAndSegmentsFromFragment(String fragment) {
+  final [category, ...segments] = fragment.split('/');
+  if (segments.length > 1 && segments.last == '') segments.removeLast();
+  return (category, segments);
+}
+
+Narrow? _interpretNarrowSegments(List<String> segments, PerAccountStore store) {
+  assert(segments.isNotEmpty);
+  assert(segments.length.isEven);
+
+  ApiNarrowStream? streamElement;
+  ApiNarrowTopic? topicElement;
+  ApiNarrowDm? dmElement;
+
+  for (var i = 0; i < segments.length; i += 2) {
+    final (operator, negated) = _parseOperator(segments[i]);
+    // TODO(#252): negated operators are not currently supported
+    // These should eventually return a SearchNarrow
+    if (negated) return null;
```

Comment:
```suggestion
    // TODO(#252): negated operators are not currently supported;
    //   they should eventually return a SearchNarrow
    if (negated) return null;
```

That is, I _think_ the intent here is that the second comment line here is elaborating on the TODO, and "these" refers to negated operators; this would be a way of making that explicit. In particular the indentation groups it into the TODO.

If "these" is intended to mean something else then I guess it'd be good to make that explicit instead :-)

Reply:
Hmm and then reading a bit more of the code around this: there are several other spots in this function where a similar comment applies, right?

I think the biggest one is probably the `_NarrowOperator.unknown` case — that's what covers plain search words, and `sender:`, and miscellaneous other search features. But also having multiple stream, topic, or DM elements; or having a DM element and a stream or topic element; and maybe some of the other `return null` cases too.

So the cleanest way to express this point may be to put a comment up somewhere like the top of the function. For example… oh, in fact we already have a discussion of this up at `parseInternalLink`. So can just add the TODO (with the handy issue-number reference) there, like:
```diff
 /// This can also return null for some valid narrow links that our Narrow
 /// type *could* accurately represent. We should try to understand these
 /// better, but some kinds will be rare, even unheard-of:
 ///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+// TODO(#252): handle all valid narrow links, returning a search narrow
 Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
```

I think that covers it. If we set out to implement that TODO, we'll naturally go into `_interpretNarrowSegments` and find all the places where it's returning null instead of a narrow, and identify whether each one corresponds to a valid narrow link and so should start returning a search narrow instead.